crates/holochain/src/conductor/conductor.rs

```diff
@@ -2409,26 +2407,29 @@ impl Conductor {
         let dna_modifiers = clone_dna.dna().modifiers.clone();
         let clone_dna_hash = clone_dna.dna_hash().to_owned();
 
-        // if DNA hash of new clone cell already exists, reject as duplicate
-        if self
-            .ribosome_store
-            .share_ref(|ribosome_store| ribosome_store.get_dna_def(&clone_dna_hash))
-            .is_some()
-        {
-            return Err(ConductorError::AppError(AppError::DuplicateDnaHash(
-                clone_dna_hash,
-            )));
-        }
-
         // add clone cell to app and instantiate resulting clone cell
         let (_, installed_clone_cell) = self
             .update_state_prime(move |mut state| {
+                let state_copy = state.clone();
                 let app = state.get_app_mut(&app_id)?;
                 let agent_key = app.role(&role_name)?.agent_key().to_owned();
-                let cell_id = CellId::new(clone_dna_hash, agent_key);
-                let clone_id = app.add_clone(&role_name, &cell_id)?;
+                let clone_cell_id = CellId::new(clone_dna_hash, agent_key);
+
+                // if cell id of new clone cell already exists, reject as duplicate
+                if state_copy
+                    .installed_apps()
+                    .iter()
+                    .flat_map(|(_, app)| app.all_cells())
+                    .any(|cell_id| *cell_id == clone_cell_id)
+                {
+                    return Err(ConductorError::AppError(AppError::DuplicateCellId(
+                        clone_cell_id,
+                    )));
+                }
```

Comment:
This is good, I wonder if we have this check when creating a normal cell too?

Reply:
We don't, no